[SPARK-24020][SQL] Sort-merge join inner range optimization

zecevicp opened the pull request:
## JIRA description

The problem we are solving is the case where you have two big tables partitioned by X column, but also sorted within partitions by Y column and you need to calculate an expensive function on the joined rows, which reduces the number of output rows (e.g. condition based on a spatial distance calculation). But you could theoretically reduce the number of joined rows for which the calculation itself is performed by using a range condition on the Y column. Something like this:

`... WHERE t1.X = t2.X AND t1.Y BETWEEN t2.Y - d AND t2.Y + d AND <function calculation...>`

However, during a sort-merge join with this range condition specified, Spark will first cross-join all the rows with the same X value and only then will it try to apply the range condition and any function calculations. This happens because, inside the generated sort-merge join (SMJ) code, these extra conditions are put in the same block with the function being calculated and there is no way to evaluate these conditions before reading all the rows to be checked into memory (into an `ExternalAppendOnlyUnsafeRowArray`). If the two tables have a large number of rows per X, this can result in a huge number of calculations and a huge number of rows in executor memory, which can be unfeasible. 

We therefore propose a change to the sort-merge join so that, when these extra conditions are specified, a queue is used instead of the `ExternalAppendOnlyUnsafeRowArray` class. This queue is then used as a moving window through the values from the right relation as the left row changes. You could call this a combination of an equi-join and a theta join; in literature it is sometimes called an “epsilon join”. We call it a "sort-merge inner range join".

This design uses much less memory (not all rows with the same values of X need to be loaded into memory at once) and requires a much lower number of comparisons (the validity of this statement depends on the actual data and conditions used). 

The optimization should be triggered automatically when an equi-join expression is present AND lower and upper range conditions on a secondary column are specified. If the tables aren't sorted by both columns, appropriate sorts should be added.

To limit the impact of this change we also propose adding a new parameter (tentatively named `spark.sql.join.smj.useInnerRangeOptimization`) which could be used to switch off the optimization entirely.

## What changes were proposed in this pull request?

The main changes are made to these classes:

`ExtractEquiJoinKeys` – a pattern that needs to be extended to be able to recognize the case where a simple range condition with lower and upper limits is used on a secondary column (a column not included in the equi-join condition). The pattern also needs to extract the information later required for code generation etc.

`ExternalAppendOnlyUnsafeRowArray` – changed so that it can function as a queue too. The rows need to be removed and added to/from the structure as the left key (X) changes, or the left secondary value (Y) changes, so the structure needs to be a queue. The class is no longer "append only", but I didn't want to change too many things.

`JoinSelection` – a strategy that uses `ExtractEquiJoinKeys` and needs to be aware of the extracted range conditions

`SortMergeJoinExec` – the main implementation of the optimization. Needs to support two code paths: 
    • when whole-stage code generation is turned off (method `doExecute`, which uses `sortMergeJoinInnerRangeScanner`) 
    • when whole-stage code generation is turned on (methods `doProduce` and `genScanner`)
`SortMergeJoinInnerRangeScanner` – implements the SMJ with inner-range optimization in the case when whole-stage codegen is turned off

## How was this patch tested?
Unit tests (`InnerRangeSuite.scala`, `JoinSuite.scala`, `ExternalAppendOnlyUnsafeRowArraySuite`)

Performance tests in `JoinBenchmark`. The tests show 8x improvement over non-optimized code. Although, it should be noted that the results depend on the exact range conditions and the calculations performed on each matched row. 
In our case, we were not able to cross-match two rather large datasets (1.2 billion rows x 800 million rows) without this optimization. With the optimization, the cross-match finishes in less than 2 minutes.

Commit:
Secondary sort Sort Merge Join optimization. Not finished yet.

Commit:
SortMergeJoin secondary sort optimization

Commit:
Sort-Merge "inner range join" (secondary sort) - code generation

Commit:
Sort-Merge "inner range join" (secondary sort) - two bug fixes - works now

Commit:
Code simplification

Commit:
Bug fix

Commit:
Scalastyle fixes

Commit:
Scalastyle fixes

Commit:
SMJ range join unit tests

Commit:
Scalastyle

Commit:
Scalastyle

Commit:
Scalastyle

Commit:
Fix generated code - dequeue method missing

Commit:
Bug fix: include other binary comparisons in range conditions match

Commit:
Test fix: sortWithinPartitions; Bug Fix: check references in rangeConditions, not columns

Commit:
Test fix

Commit:
Test fix

Commit:
Fix required child ordering for inner range queries

Commit:
Parameter for turning off inner range optimization

Commit:
Scala style

Commit:
Bug fix - NPE when inner range optimization turned off

Commit:
Adding test case when inner range optimization is turned off

Commit:
Stala style

Commit:
Stala style

Commit:
Remove range condition extraction when inner range join optimization is disabled

Commit:
Scala style

Commit:
Unit test fix

Commit:
Unit test fix

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #89592 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89592/testReport)** for PR 21109 at commit [`4c6a726`](https://github.com/apache/spark/commit/4c6a726bac71796d7210cfde9cf762dfdc38d165).

SparkQA commented:
**[Test build #89592 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89592/testReport)** for PR 21109 at commit [`4c6a726`](https://github.com/apache/spark/commit/4c6a726bac71796d7210cfde9cf762dfdc38d165).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89592/
Test FAILed.

zecevicp commented:
Can you please restart the build? It took 5 hours for some reason, but the "inner range join" tests completed successfully.

tedyu commented:
retest this please

SparkQA commented:
**[Test build #89641 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89641/testReport)** for PR 21109 at commit [`4c6a726`](https://github.com/apache/spark/commit/4c6a726bac71796d7210cfde9cf762dfdc38d165).

SparkQA commented:
**[Test build #89641 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89641/testReport)** for PR 21109 at commit [`4c6a726`](https://github.com/apache/spark/commit/4c6a726bac71796d7210cfde9cf762dfdc38d165).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89641/
Test PASSed.

maropu commented:
Can you put performance numbers w/ this pr? Also, you'd be better to add benchmark code in `JoinBenchmark`.

Commit:
- Turning off inner range optimization when whole stage code generation is turned off.
- Debugging
- comments

Commit:
Switch off inner range optimization when whole stage codegen is off.

Commit:
SMJ inner range optimization benchmarks

SparkQA commented:
**[Test build #89934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89934/testReport)** for PR 21109 at commit [`fbec452`](https://github.com/apache/spark/commit/fbec452e6376d35c8001544d96fe42831e790f9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89934/
Test PASSed.

viirya commented:
Thanks for working on this.

Based on the description on JIRA, I think the main cause of the bad performance is re-calculation an expensive function on matches rows. With the added benchmark, I adjust the order of conditions so the expensive UDF is put at the end of predicate. Below is the results. The first one is original benchmark. The second is the one with UDF at the end of predicate.


```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16 on Linux 4.9.87-linuxkit-aufs
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
sort merge inner range join:             Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
sort merge inner range join wholestage off      6913 / 6964          0.0     1080112.4       1.0X
sort merge inner range join wholestage on      2094 / 2224          0.0      327217.4       3.3X
```

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_152-b16 on Linux 4.9.87-linuxkit-aufs
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
sort merge inner range join:             Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
sort merge inner range join wholestage off       675 /  704          0.0      105493.9       1.0X
sort merge inner range join wholestage on       374 /  398          0.0       58359.6       1.8X
```

It can be easily improved because short-circuit evaluation of predicate. This can be applied to also other conditions other than just range comparison. So I'm thinking if we need a way to give a hint to Spark to adjust the order of expression for an expensive one like UDF.

Commit:
Removing "expensive function" from the SMJ inner range optimization benchmark.

SparkQA commented:
**[Test build #89961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89961/testReport)** for PR 21109 at commit [`e6e6628`](https://github.com/apache/spark/commit/e6e6628bf3d63e0486c2ba90c03712aa0eade013).

zecevicp commented:
Hey Liang-Chi, thanks for looking into this.
Yes, the problem can be circumvented by changing the join condition as you describe, but only in the benchmark case, because my "expensive function" was a bit misleading. 
The problem is not in the function itself, but in the number of rows that are checked for each pair of matching equi-join keys. 
I changed the benchmark test case now so to better demonstrate this. I completely removed the expensive function and I'm only doing a count on the matched rows. The results are the following.
Without the optimization:
```
AMD EPYC 7401 24-Core Processor
sort merge join:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------
sort merge join wholestage off            30956 / 31374          0.0       75575.5       1.0X
sort merge join wholestage on             10864 / 11043          0.0       26523.6       2.8X
```
With the optimization:
```
AMD EPYC 7401 24-Core Processor
sort merge join:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------
sort merge join wholestage off            30734 / 31135          0.0       75035.2       1.0X
sort merge join wholestage on                959 / 1040          0.4        2341.3      32.0X
```
This shows a 10x improvement over the non-optimized case (as I already said, this depends on the range condition, number of matched rows, the calculated function, etc.).

Regarding your second question as to why is the "wholestage off" case in the optimized version so slow, that is because the optimization is turned off when the wholestage code generation is turned off.
And that is simply because it was too hard to debug it and I figured the wholestage generation is on by default, so I'm guessing (and hoping) that it would not be too hard of a requirement to have to turn wholestage codegen on if you want to use this optimization.

mgaido91 commented:
@zecevicp wholestage codegen now is turned on by default only if we have few columns (less than 100). This can be false in many real use-cases. Is there any specific reason why this optimization cannot be applied to the non-wholestage codegen case? If not, I think it is worth to consider also this case.

zecevicp commented:
Hi, Gaido, thanks for the comment. As I said, it was difficult to debug it and I didn't have time. We might open a different ticket for the non-wholestage codegen case, once this is merged?

mgaido91 commented:
Why do you say that it is difficult to debug? What was difficult?

zecevicp commented:
The code path with the optimization but without wholegen code generation gives wrong results. And I haven't been able to figure out where is the bug. I spent several hours at this again today. I've been using this code as an example:
```
spark.conf.set("spark.sql.codegen.wholeStage", value = false)
spark.conf.set("spark.sql.shuffle.partitions", "1")
val df1 = sc.parallelize(1 to 2).cartesian(sc.parallelize(1 to 10)).toDF("col1a", "col1b")
val df2 = sc.parallelize(1 to 2).cartesian(sc.parallelize(1 to 10)).toDF("col2a", "col2b")
val res = df1.join(df2, 'col1a === 'col2a and ('col1b < 'col2b + 3) and ('col1b > 'col2b - 3))
res.count
```
With wholeStage off this gives 65 as a result and 88 if wholeStage is on (with the optimization turned on). Without the optimization it is always 88, which is correct.
So, if someone can figure out where the bug is, great. If not, maybe it's best to just turn the optimization off for the non-wholegen case (for now).

maropu commented:
Since I feel this is a limited case, I'm not certainly sure this optimization needs to be handled in smj. For spatial or temporal use cases, is it not enough to add dummy join keys to split tasks into pieces for workaround?
```
'col1a === 'col2a and 'col1dummyKey === 'col2dummyKey and ('col1b < 'col2b + 3) and ('col1b > 'col2b - 3)
```
Btw, can you fix this issue by more simpler code change?  (I'm not sure this big change pays the performance gain...)

maropu commented:
@viirya I think predicate reordering based on cost estimation  and others is an interesting topic in optimizer. But, IMO the topic is not directly related to this pr (I'm not 100% sure though), it'd be better to file another jira to keep the discussion (and the benchmark result you reported)? I couldn't find a jira entry for the topic.

Commit:
SMJ inner range optimization with wholeStage codegen turned off - code + tests

Commit:
Unit test fix. Benchmark results update.

SparkQA commented:
**[Test build #90475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90475/testReport)** for PR 21109 at commit [`535d0d6`](https://github.com/apache/spark/commit/535d0d63b33ef320fded52c18b2716a1333255ce).

SparkQA commented:
**[Test build #90475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90475/testReport)** for PR 21109 at commit [`535d0d6`](https://github.com/apache/spark/commit/535d0d63b33ef320fded52c18b2716a1333255ce).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90475/
Test FAILed.

zecevicp commented:
I managed to fix the code path that is executed when the wholestage codegen is turned off. Now both code paths give the same results and have the optimization implemented. I also changed the tests in the `InnerJoinSuite` class so that they are run with both wholestage turned on off and on (which wasn't the case so far).
I updated the benchmark results in `JoinBenchmark`. The results are now the following.
Without inner range optimization:
```
sort merge join:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------
sort merge join wholestage off            25226 / 25244          0.0       61585.9       1.0X
sort merge join wholestage on              8581 / 8983          0.0       20948.6       2.9X
```
With inner range optimization:
```
sort merge join:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------
sort merge join wholestage off              1194 / 1212          0.3        2915.2       1.0X
sort merge join wholestage on                814 /  867          0.5        1988.4       1.5X
```
So, there is 10x improvement for wholestage ON case and 21x improvement for wholestage OFF case.

I believe this is now ready to be merged, which would greatly help us in our projects.

Commit:
Scalastyle for comments

SparkQA commented:
**[Test build #90477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90477/testReport)** for PR 21109 at commit [`6dc9000`](https://github.com/apache/spark/commit/6dc90002602d7790d2ab4a9d4946f61c4c8c078e).

SparkQA commented:
**[Test build #90477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90477/testReport)** for PR 21109 at commit [`6dc9000`](https://github.com/apache/spark/commit/6dc90002602d7790d2ab4a9d4946f61c4c8c078e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90477/
Test PASSed.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
         val (leftKeys, rightKeys) = joinKeys.unzip
-        logDebug(s"leftKeys:$leftKeys | rightKeys:$rightKeys")
-        Some((joinType, leftKeys, rightKeys, otherPredicates.reduceOption(And), left, right))
+
```
nit: unnecessary empty line

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+        // and which can be used for secondary sort optimizations.
+        val rangePreds: mutable.Set[Expression] = mutable.Set.empty
+        var rangeConditions: Seq[BinaryComparison] =
+          if (SQLConf.get.useSmjInnerRangeOptimization) { // && SQLConf.get.wholeStageEnabled) {
```
remove the comment please

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+        // of the two tables being joined,
+        // which are not used in the equijoin expressions,
+        // and which can be used for secondary sort optimizations.
+        val rangePreds: mutable.Set[Expression] = mutable.Set.empty
```
why do we need this? Can't we just use `rangeConditions`?

mgaido91 replied:
I see that you are using it later but can't you use `rangeConditions` instead? they seem duplicates...they contain the same data IIUC

zecevicp replied:
`rangeConditions` contain "vice-versa" conditions in case left and right plans need to be switched.

mgaido91 replied:
I see... maybe we can make this more clear in the comments then, what do you think?

zecevicp replied:
Yes, will do

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
   }
+
+  private def isValidRangeCondition(l : Expression, r : Expression,
+                                    left : LogicalPlan, right : LogicalPlan,
```
nit: bad indent

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+                                    joinKeys : Seq[(Expression, Expression)]) = {
+    val (lattrs, rattrs) = (l.references.toSeq, r.references.toSeq)
+    if(lattrs.size != 1 || rattrs.size != 1) {
+      "none"
```
why not returning an `Option` and make this a `None`?

zecevicp replied:
Seemed simpler. I can change it if you insist.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
     case _ => None
   }
+
+  private def isValidRangeCondition(l : Expression, r : Expression,
```
can we add the return type and add some description to this method? Moreover the name doesn't seem correct, since I'd expect it returning a boolean from its name.

zecevicp replied:
That's true..

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+      val equiset = joinKeys.filter{ case (ljk : Expression, rjk : Expression) =>
+        ljk.references.toSeq.contains(lattrs(0)) && rjk.references.toSeq.contains(rattrs(0)) }
+      if (equiset.isEmpty) {
+        "asis"
```
can we return something more meaningful than this string? Maybe a `Option(Bool)` in enough which tells whether to reverse or not.

zecevicp replied:
OK, makes sense.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+      "none"
+    }
+    else if (canEvaluate(l, left) && canEvaluate(r, right)) {
+      val equiset = joinKeys.filter{ case (ljk : Expression, rjk : Expression) =>
```
what about using exists?

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:
```diff
@@ -434,18 +511,139 @@ case class SortMergeJoinExec(
+      val rightRngTmpKeyVars = createJoinKey(ctx, rightTmpRow,
+        rightUpperKeys.slice(0, 1), right.output)
+      val rightRngTmpKeyVarsDecl = rightRngTmpKeyVars.map(_.code).mkString("\n")
+      rightRngTmpKeyVars.foreach(_.code = "")
```
why are we doing this?

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:
```diff
@@ -434,18 +511,139 @@ case class SortMergeJoinExec(
+        rightLowerKeys.size == 0 || rightUpperKeys.size == 0) {
+      ctx.addNewFunction("dequeueUntilUpperConditionHolds",
+        "private void dequeueUntilUpperConditionHolds() { }",
+        inlineToOuterClass = true)
```
do we really need this? can't we just use the returned name?

mgaido91 replied:
can we avoid to use `inlineToOuterClass = true`? I think we can avoid doing that.

zecevicp replied:
Ah, OK. I'll investigate.

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:
```diff
@@ -840,6 +1049,241 @@ private[joins] class SortMergeJoinScanner(
+ * @param spillThreshold Threshold for number of rows to be spilled by internal buffer
+ */
+private[joins] class SortMergeJoinInnerRangeScanner(
+                                           streamedKeyGenerator: Projection,
```
nit: indent

mgaido91 reviewed (commented):
I just left some comments after a quick pass, but I am a bit concerned about the amount of changes we need for this patch. May you please check if you can reduce the amount of code changes?

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -97,13 +100,13 @@ object PhysicalOperation extends PredicateHelper {
 
   def unapply(plan: LogicalPlan): Option[ReturnType] = plan match {
     case join @ Join(left, right, joinType, condition) =>
-      logDebug(s"Considering join on: $condition")
```
I don't think so, this is the diff with the master, it was added in 2014....

zecevicp replied:
OK, you could be right. I'll put it back.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +134,101 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+                }
+              case _ => None
+            }
+          }
```
nit: `} else {`?

Commit:
Code changes based on review comments.

SparkQA commented:
**[Test build #90642 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90642/testReport)** for PR 21109 at commit [`48b1c0e`](https://github.com/apache/spark/commit/48b1c0e73d3db1c406b719a2343a9c395627bea2).

zecevicp commented:
@mgaido91 Regarding the amount of code, maybe you can suggest how to reduce it? Because I don't see a way...
I think the code is well contained (mostly in separate new classes) and is not contaminating the existing codebase. There is a simple switch (parameter) that turns off the whole thing.
Some code also went into the use cases which contribute to the amount of changes.
I believe that the potentially significant speedup justifies the line count.

mgaido91 commented:
@zecevicp for instance do we really need `InMemoryUnsafeRowQueue`? why `ExternalAppendOnlyUnsafeRowArray` is not ok?

zecevicp commented:
Well, that is the essence of the contribution: to have a moving window over the data, instead of a fixed block (per equi-join match). To implement a moving window you need something like a queue.

SparkQA commented:
**[Test build #90642 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90642/testReport)** for PR 21109 at commit [`48b1c0e`](https://github.com/apache/spark/commit/48b1c0e73d3db1c406b719a2343a9c395627bea2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90642/
Test PASSed.

srowen reviewed (commented):
Just some tiny code comments. I don't feel qualified to merge this, but would instead get a nod from at least one of people like @marmbrus @cloud-fan @gatorsmile or others who know the SQL planning code.

zecevicp commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1205,6 +1205,19 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val USE_SMJ_INNER_RANGE_OPTIMIZATION =
```
It's just a safety valve. In case there are some queries that I don't foresee now where this could get in the way.

zecevicp commented on sql/core/src/main/scala/org/apache/spark/sql/execution/InMemoryUnsafeRowQueue.scala:
```diff
@@ -0,0 +1,183 @@
+ *   excessive disk writes. This may lead to a performance regression compared to the normal case
+ *   of using an [[ArrayBuffer]] or [[Array]].
+ */
+private[sql] class InMemoryUnsafeRowQueue(
```
A queue is needed here because it's a moving window instead of a fixed block of rows. Maybe I missed an existing class that could do this easily so I'll take another look. But, I believe any alternative would indeed be messier.

zecevicp commented on sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala:
```diff
@@ -117,101 +131,170 @@ class InnerJoinSuite extends SparkPlanTest with SharedSQLContext {
+
+    // Disabling these because the code would never follow this path in case of a inner range join
+    if (!expectRangeJoin) {
+      var counter = 1
```
OK, will do that.

zecevicp commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -131,13 +135,100 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
+        // which are not used in the equijoin expressions,
+        // and which can be used for secondary sort optimizations.
+        // rangePreds will contain the original expressions to be filtered out later.
+        val rangePreds: mutable.Set[Expression] = mutable.Set.empty
```
I think you're right.

kiszk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/InMemoryUnsafeRowQueue.scala:
```diff
@@ -0,0 +1,183 @@
+    null
+  }
+
+//  private var spillableArray: UnsafeExternalSorter = _
```
nit: Is this comment necessary?

zecevicp replied:
No, it's not. Thank you

Commit:
Code review changes

Commit:
Removing exception when numRowsInMemoryBufferThreshold is reached in InMemoryUnsafeRowQueue class.

SparkQA commented:
**[Test build #91533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91533/testReport)** for PR 21109 at commit [`9a0b2ab`](https://github.com/apache/spark/commit/9a0b2abf482a51c1e9901ce5297c9ca1b7961765).

SparkQA commented:
**[Test build #91534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91534/testReport)** for PR 21109 at commit [`3fa690f`](https://github.com/apache/spark/commit/3fa690faf4e9a0b7d8eb2a5854ebb6a854a44d2a).

Commit:
Scala style

SparkQA commented:
**[Test build #91534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91534/testReport)** for PR 21109 at commit [`3fa690f`](https://github.com/apache/spark/commit/3fa690faf4e9a0b7d8eb2a5854ebb6a854a44d2a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91534/
Test FAILed.

zecevicp commented:
Retest this please

zecevicp commented:
@tedyu @gatorsmile Can you please trigger the build again? It failed for some unrelated reason. Thanks

zecevicp commented:
@davies pointed out that `InMemoryUnsafeRowQueue` was throwing an exception when `numRowsInMemoryBufferThreshold` was reached. That is now fixed and only a warning is logged.

cloud-fan commented:
do we have a design doc? I have a couple of high-level questions:
1. what if the range is big and the queue OOM?
2. can't we apply it with cartesian join if there is no equal condition? and what about broadcast join?
3. if the equal join key is skewed, then we kind of doing this sort-merge-range-join in a single thread, even if the range join key is not skewed.

zecevicp commented:
There's no design doc. I didn't feel the change was big enough to warrant one.

1. Currently there is no spill-over to disk. If the range is too big, users can switch this off and use the much slower SMJ version, without an OOM. Implementing spill-over doesn't look trivial because it's more dynamic than the original version. It's not clear how to implement that. Maybe we can add that in the future, once we figure it out?
2. This whole optimization doesn't apply when there is no equal condition.
3. I didn't understand this case you're describing. Can you elaborate, please? Either way, only one pass through the data is needed, skewed or not skewed.

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #91760 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91760/testReport)** for PR 21109 at commit [`5c64b55`](https://github.com/apache/spark/commit/5c64b55766630ffec33ac3c82ce10703dc3d526c).

Commit:
Unneeded import

SparkQA commented:
**[Test build #91760 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91760/testReport)** for PR 21109 at commit [`5c64b55`](https://github.com/apache/spark/commit/5c64b55766630ffec33ac3c82ce10703dc3d526c).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91760/
Test FAILed.

Commit:
A dot

SparkQA commented:
**[Test build #91785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91785/testReport)** for PR 21109 at commit [`fb99390`](https://github.com/apache/spark/commit/fb99390a4baf62ce953be5a78c552adae98c1ffa).

SparkQA commented:
**[Test build #91785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91785/testReport)** for PR 21109 at commit [`fb99390`](https://github.com/apache/spark/commit/fb99390a4baf62ce953be5a78c552adae98c1ffa).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91785/
Test FAILed.

Commit:
A dot

SparkQA commented:
**[Test build #91824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91824/testReport)** for PR 21109 at commit [`82c194a`](https://github.com/apache/spark/commit/82c194a8a03b6cc028de303fbc07c68d6078cc2b).

SparkQA commented:
**[Test build #91824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91824/testReport)** for PR 21109 at commit [`82c194a`](https://github.com/apache/spark/commit/82c194a8a03b6cc028de303fbc07c68d6078cc2b).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91824/
Test FAILed.

Commit:
A dot

SparkQA commented:
**[Test build #91900 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91900/testReport)** for PR 21109 at commit [`6d72fe0`](https://github.com/apache/spark/commit/6d72fe0466f210e001a67829e0f42379abc7e4f0).

SparkQA commented:
**[Test build #91900 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91900/testReport)** for PR 21109 at commit [`6d72fe0`](https://github.com/apache/spark/commit/6d72fe0466f210e001a67829e0f42379abc7e4f0).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91900/
Test FAILed.

SparkQA commented:
**[Test build #91912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91912/testReport)** for PR 21109 at commit [`3e9b3e2`](https://github.com/apache/spark/commit/3e9b3e2be380d3973cb559ce88b3fff588f3dd3e).

SparkQA commented:
**[Test build #91912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91912/testReport)** for PR 21109 at commit [`3e9b3e2`](https://github.com/apache/spark/commit/3e9b3e2be380d3973cb559ce88b3fff588f3dd3e).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91912/
Test FAILed.

SparkQA commented:
**[Test build #91913 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91913/testReport)** for PR 21109 at commit [`746fad3`](https://github.com/apache/spark/commit/746fad3f9caf5d8d43f91b66c45091ef52411ce2).

SparkQA commented:
**[Test build #91913 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91913/testReport)** for PR 21109 at commit [`746fad3`](https://github.com/apache/spark/commit/746fad3f9caf5d8d43f91b66c45091ef52411ce2).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91913/
Test FAILed.

SparkQA commented:
**[Test build #91917 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91917/testReport)** for PR 21109 at commit [`8eaab13`](https://github.com/apache/spark/commit/8eaab130b9eb6b9f3224c323165b6dc2b175f8ac).

SparkQA commented:
**[Test build #91917 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91917/testReport)** for PR 21109 at commit [`8eaab13`](https://github.com/apache/spark/commit/8eaab130b9eb6b9f3224c323165b6dc2b175f8ac).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91917/
Test FAILed.

Commit:
A dot

SparkQA commented:
**[Test build #92023 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92023/testReport)** for PR 21109 at commit [`66d7cbf`](https://github.com/apache/spark/commit/66d7cbf76ec0170bd8e78d4b936e6c7650998f34).

SparkQA commented:
**[Test build #92023 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92023/testReport)** for PR 21109 at commit [`66d7cbf`](https://github.com/apache/spark/commit/66d7cbf76ec0170bd8e78d4b936e6c7650998f34).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92023/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #92038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92038/testReport)** for PR 21109 at commit [`66d7cbf`](https://github.com/apache/spark/commit/66d7cbf76ec0170bd8e78d4b936e6c7650998f34).

SparkQA commented:
**[Test build #92038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92038/testReport)** for PR 21109 at commit [`66d7cbf`](https://github.com/apache/spark/commit/66d7cbf76ec0170bd8e78d4b936e6c7650998f34).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92038/
Test FAILed.

SparkQA commented:
**[Test build #92071 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92071/testReport)** for PR 21109 at commit [`3ccc292`](https://github.com/apache/spark/commit/3ccc2929dfd828e333d4deaacdb993e9fc7e5f28).

SparkQA commented:
**[Test build #92071 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92071/testReport)** for PR 21109 at commit [`3ccc292`](https://github.com/apache/spark/commit/3ccc2929dfd828e333d4deaacdb993e9fc7e5f28).
 * This patch **fails due to an unknown error code, -9**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92071/
Test FAILed.

Commit:
A dot

SparkQA commented:
**[Test build #92089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92089/testReport)** for PR 21109 at commit [`9889ba1`](https://github.com/apache/spark/commit/9889ba1ddbf0bfcb4d48b890634b6389ac4bd535).

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92089/
Test FAILed.

SparkQA commented:
**[Test build #92247: Deflake Build #92089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92247/testReport)** for PR 21109 at commit [`9889ba1`](https://github.com/apache/spark/commit/9889ba1ddbf0bfcb4d48b890634b6389ac4bd535).

SparkQA commented:
**[Test build #92247: Deflake Build #92089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92247/testReport)** for PR 21109 at commit [`9889ba1`](https://github.com/apache/spark/commit/9889ba1ddbf0bfcb4d48b890634b6389ac4bd535).
 * This patch **fails due to an unknown error code, -9**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #92435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92435/testReport)** for PR 21109 at commit [`48c3929`](https://github.com/apache/spark/commit/48c392906488167141355ac4e0ef41c449357280).

SparkQA commented:
**[Test build #92435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92435/testReport)** for PR 21109 at commit [`48c3929`](https://github.com/apache/spark/commit/48c392906488167141355ac4e0ef41c449357280).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92435/
Test FAILed.

SparkQA commented:
**[Test build #92436 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92436/testReport)** for PR 21109 at commit [`497aa52`](https://github.com/apache/spark/commit/497aa52c9a0e15b1fdf0674c9b9510deef9e5dc1).

ctslater committed:
Fixes for some rebase issues.

SparkQA commented:
**[Test build #92441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92441/testReport)** for PR 21109 at commit [`a2a5f82`](https://github.com/apache/spark/commit/a2a5f82c377402348a82e5db5587504d39f5a894).

SparkQA commented:
**[Test build #92441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92441/testReport)** for PR 21109 at commit [`a2a5f82`](https://github.com/apache/spark/commit/a2a5f82c377402348a82e5db5587504d39f5a894).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92441/
Test FAILed.

zecevicp committed:
Merge with upstream

SparkQA commented:
**[Test build #92467 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92467/testReport)** for PR 21109 at commit [`7f7ab25`](https://github.com/apache/spark/commit/7f7ab257bc14f5b529de6c22b45f00bc2724ab20).

SparkQA commented:
**[Test build #92467 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92467/testReport)** for PR 21109 at commit [`7f7ab25`](https://github.com/apache/spark/commit/7f7ab257bc14f5b529de6c22b45f00bc2724ab20).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92467/
Test FAILed.

zecevicp committed:
Merge with upstream

SparkQA commented:
**[Test build #92472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92472/testReport)** for PR 21109 at commit [`39247ba`](https://github.com/apache/spark/commit/39247bac0de645aa959cc7fd11a27e36532181a5).

SparkQA commented:
**[Test build #92472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92472/testReport)** for PR 21109 at commit [`39247ba`](https://github.com/apache/spark/commit/39247bac0de645aa959cc7fd11a27e36532181a5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92472/
Test PASSed.

cloud-fan commented:
I think range join is a very important feature, but the current implementation is not robust enough(no spilling to disk). Have you explored other simpler solutions? I vaguely remember that some papers describe a way to rewrite range join to equal-join during the logical phase.

zecevicp commented:
Implementing spilling over seems a lot of work because this is a queue. If data is spilled over to disk and you need to pop from the queue, it is not clear to me what is the best way to do that. Do you spill over only one part of the queue (so that you can add or pop more efficiently)? Which part (the beginning or the end)? Or maybe the middle? What is the threshold to bring it back to memory from disk? And other similar questions...
But I think it can be expected that much less memory will be consumed by the queue, compared to the original `ExternalAppendOnlyUnsafeRowArray`, because the queue's purpose IS to reduce the number of rows in memory, so spill-over would rarely be needed (that would depend, of course, to the user's range condition). 
That's why implementing spilling over doesn't seem critical to me. I can try and implement it, if everybody thinks it's really needed, but as I said, it's not clear (to me) what would be the best approach.

Regarding the second point, this is not an ordinary range join, but an equi-join with a secondary range condition.

cloud-fan commented:
> Regarding the second point, this is not an ordinary range join, but an equi-join with a secondary range condition.

"an equi-join with a secondary range condition" is a restriction isn't it? IIRC the logical rewrite approach can work for range join without equi-join, which has broader use cases. e.g. turning a cartisian join into hash join/SMJ.

I took a quick look at this patch, it's really bulky, and not safe to be turned on by default without the spill ability. And more importantly, it can't work for range join without equi-condition, which I think should be a common use case.

Can you investigate the logical rewrite approach? I'll attach some related papers if I find them.

zecevicp commented:
That is a restriction if you regard this as an implementation of a range join, which it is not. This is an *optimization of an equi join*. The intention never was for it to work as a range join. 
I can investigate the logical rewrite approach, but as a part of a different JIRA ticket, one which would actually be related to a range join.
Thanks for looking into this, though.

zecevicp committed:
SMJ inner range spill over implementation and tests

zecevicp committed:
External unsafe row dequeue test extension

SparkQA commented:
**[Test build #94630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94630/testReport)** for PR 21109 at commit [`62a8071`](https://github.com/apache/spark/commit/62a807168db9522673c194fbfa8c326c8688da82).

zecevicp committed:
A dot

SparkQA commented:
**[Test build #94631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94631/testReport)** for PR 21109 at commit [`af59b8a`](https://github.com/apache/spark/commit/af59b8a285ffe968d87cdf1bf9758b142e0bcfd4).

SparkQA commented:
**[Test build #94630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94630/testReport)** for PR 21109 at commit [`62a8071`](https://github.com/apache/spark/commit/62a807168db9522673c194fbfa8c326c8688da82).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94630/
Test FAILed.

SparkQA commented:
**[Test build #94631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94631/testReport)** for PR 21109 at commit [`af59b8a`](https://github.com/apache/spark/commit/af59b8a285ffe968d87cdf1bf9758b142e0bcfd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94631/
Test PASSed.

cloud-fan commented:
What's the advantage of this feature when Spark can rewrite range join to equal join logically?

BTW I also hesitate to merge such a big patch to the SQL engine since we are close to code freeze.

zecevicp commented:
Sorry, I don't quite understand your question. This already applies (only) to equijoins if there are additional range conditions on secondary columns. So if Spark rewrites those range conditions (?), and you end up with two equijoins (doesn't sound like a realistic scenario), then this doesn't apply at all.
But if you have an equi-join which cannot be performed because you need to match a huge number of rows, and you can narrow down the search window using range conditions, then the advantage is that this makes it feasible and/or much, much faster.

Regarding the second point, I don't want to tell you what to do, but on my part I can say that this has been tested with unit tests and on real, large datasets and I believe it should be safe to merge. But it can also wait for 2.5/3.0...

ctslater commented:
@cloud-fan, could you point us at the papers that describe rewriting range join into equi-join? That would be a very convenient solution, but I've been digging through the literature and haven't found anything showing how to do that. All the experiments I've tried result in cartesian joins.

SparkQA commented:
**[Test build #95807 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95807/testReport)** for PR 21109 at commit [`0a5c8de`](https://github.com/apache/spark/commit/0a5c8de7769315934712bf853401c332dd747a6e).

SparkQA commented:
**[Test build #95807 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95807/testReport)** for PR 21109 at commit [`0a5c8de`](https://github.com/apache/spark/commit/0a5c8de7769315934712bf853401c332dd747a6e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95807/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

zecevicp committed:
Merge branch 'master' into branch-pz-smj

SparkQA commented:
**[Test build #99907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99907/testReport)** for PR 21109 at commit [`07ff4d3`](https://github.com/apache/spark/commit/07ff4d3a69967e13438dd8bd3e4130bf23b65c7d).

SparkQA commented:
**[Test build #99907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99907/testReport)** for PR 21109 at commit [`07ff4d3`](https://github.com/apache/spark/commit/07ff4d3a69967e13438dd8bd3e4130bf23b65c7d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99907/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?
